commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/filesystem/FileSystemStorage.java

```diff
@@ -73,6 +77,9 @@ public InputStream fetch(final String key, final int from, final int to) throws
         }
 
         final Path path = fsRoot.resolve(key);
+        if (!Files.exists(path)) {
+            throw new KeyNotFoundException(this, key);
+        }
```

Comment:
I think that instead of this check, we'd better catch `NoSuchFileException` on `Files.newInputStream`: this eliminates this tiny but still possible window of the file disappearing after the check and before the opening attempt or some strange subtle disparity between the existence checks in `Files.exists` and `Files.newInputStream`.